src/Logger.cpp

```diff
@@ -0,0 +1,458 @@
+    if (!Files.empty()) {
+        for (std::vector<FileRec *>::iterator i = Files.begin(); i != Files.end(); i++) {
+            if (*i != nullptr) {
+                if ((*i)->open)
+                    fclose((*i)->file_stream);
+                delete *i;
+            }
+        }
+    }
+}
+
+// -------------------------------------------------------------
+// --- Helper
+// -------------------------------------------------------------
+
+struct Logger::Helper {
+    static std::string build_message(
+            SourceRec *rec,
+            std::string &prepend,
+            const std::string &func,
+            const std::string &file,
+            const int &line,
+            std::string what) {
+
+        std::string message;
+        if (rec->show_source_category) {
+            ((message = "(") += prepend) += ") ";
+        }
+        if (rec->show_thread_id)
+            message += thread_id;
+        if (rec->show_funct_line && !rec->funct_line_last) {
+            message.append(func).append("():").append(file).append(":").append(std::to_string(line)).append(" ");
+        }
+        message += what;
+        if (rec->show_funct_line && rec->funct_line_last) {
+            message.append(" ").append(func).append("():").append(file).append(":").append(std::to_string(line));
+        }
+        return message;
+    }
+};
+
+bool FileRec::write(std::string &msg) {
+    if (file_stream == nullptr) {
+        std::cerr << "file_stream is null" << std::endl;
+    } else {
+//        std::cerr << "file_stream is not null" << std::endl;
+        if (fprintf(file_stream, "%s\n", msg.c_str()) < 0) {
+            //std::cerr << "log write to " << filename << " failed";
+            return false;
+        }
+//        std::cerr << "log write to " << filename << " OK";
+    }
+    return true;
+}
+bool FileRec::flush() {
+    if(file_stream == nullptr)
+        return false;
+    fflush(file_stream);
+    return true;
+}
+
+bool FileRec::rotate() {   // this must only be called by a single thread which controls all output to this file
+    if(file_stream == nullptr)
+        return false;
+    std::string rfn = filename;
+    mode_t old_umask;
+    old_umask = umask(S_IWGRP | S_IWOTH);
+    rfn += ".old";
+    if(link(filename.c_str(), rfn.c_str()) == 0) {
+        unlink(filename.c_str());
+        fclose(file_stream);
+        file_stream = fopen(filename.c_str(),"a");
+        umask(old_umask);
+        return true;
+    }
+    umask(old_umask);
+    return false;
+
+}
+
+// -------------------------------------------------------------
+// --- static Functions
+// -------------------------------------------------------------
+
+LoggerSource Logger::string2source(std::string source) {
+    for (int i = 0; i < static_cast<int>(LoggerSource::__Max_Value); i++) {
+        if (source == Sources[i]) return static_cast<LoggerSource>(i);
+    }
+    return LoggerSource::none;
+}
+
+LoggerDestination Logger::string2dest(std::string destination) {
```

Comment:
Codacy found an issue: [Function parameter 'destination' should be passed by const reference.](https://app.codacy.com/manual/numsys/e2guardian/pullRequest?prid=6238852)